23_Arithmetic2UnitTest/app_resources/testWorkgroup.comp.hlsl

```diff
@@ -0,0 +1,194 @@
+        [unroll]
+        for (uint32_t idx = 0; idx < PreloadedDataCount; idx++)
+            preloaded[idx] = vk::RawBufferLoad<dtype_t>(pc.inputBufAddress + (workgroupOffset + idx * Config::WorkgroupSize + nbl::hlsl::workgroup::SubgroupContiguousIndex()) * sizeof(dtype_t));
+    }
+    void unload()
+    {
+        const uint32_t workgroupOffset = nbl::hlsl::glsl::gl_WorkGroupID().x * Config::VirtualWorkgroupSize;
+        uint64_t outputBufAddr = vk::RawBufferLoad<uint64_t>(pc.outputAddressBufAddress + Binop::BindingIndex * sizeof(uint64_t));
+        [unroll]
+        for (uint32_t idx = 0; idx < PreloadedDataCount; idx++)
+            vk::RawBufferStore<dtype_t>(outputBufAddr + sizeof(uint32_t) + sizeof(dtype_t) * (workgroupOffset + idx * Config::WorkgroupSize + nbl::hlsl::workgroup::SubgroupContiguousIndex()), preloaded[idx], sizeof(uint32_t));
+    }
+
+    void workgroupExecutionAndMemoryBarrier()
+    {
+        nbl::hlsl::glsl::barrier();
+        //nbl::hlsl::glsl::memoryBarrierShared(); implied by the above
+    }
+
+    dtype_t preloaded[PreloadedDataCount];
+};
+
+static ScratchProxy arithmeticAccessor;
+
+template<class Binop, class device_capabilities>
+struct operation_t
+{
+    using binop_base_t = typename Binop::base_t;
+    using otype_t = typename Binop::type_t;
+
+    // workgroup reduction returns the value of the reduction
+    // workgroup scans do no return anything, but use the data accessor to do the storing directly
+#if IS_REDUCTION
+    void operator()()
+    {
+        PreloadedDataProxy<config_t,Binop> dataAccessor;
+        dataAccessor.preload();
+        otype_t value = nbl::hlsl::OPERATION<config_t,binop_base_t,device_capabilities>::template __call<PreloadedDataProxy<config_t,Binop>, ScratchProxy>(dataAccessor,arithmeticAccessor);
+        // we barrier before because we alias the accessors for Binop
+        arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
+
+        [unroll]
+        for (uint32_t i = 0; i < PreloadedDataProxy<config_t,Binop>::PreloadedDataCount; i++)
+            dataAccessor.preloaded[i] = value;
+        dataAccessor.unload();
+    }
+#else
+    void operator()()
+    {
+        PreloadedDataProxy<config_t,Binop> dataAccessor;
+        dataAccessor.preload();
+        nbl::hlsl::OPERATION<config_t,binop_base_t,device_capabilities>::template __call<PreloadedDataProxy<config_t,Binop>, ScratchProxy>(dataAccessor,arithmeticAccessor);
+        // we barrier before because we alias the accessors for Binop
+        arithmeticAccessor.workgroupExecutionAndMemoryBarrier();
+        dataAccessor.unload();
+    }
+#endif
+};
+
+
+template<class Binop>
+static void subtest(NBL_CONST_REF_ARG(type_t) sourceVal)
+{
+    uint64_t outputBufAddr = vk::RawBufferLoad<uint64_t>(pc.outputAddressBufAddress + Binop::BindingIndex * sizeof(uint64_t));
+    if (globalIndex()==0u)
+        vk::RawBufferStore<uint32_t>(outputBufAddr, nbl::hlsl::glsl::gl_SubgroupSize());
```

Comment:
we can assert it equals what its supposed to equal and we can remove all the logic dealing with that one special element at the start